options/append.go

```diff
@@ -0,0 +1,25 @@
+package options
+
+import (
+	stream_revision "github.com/EventStore/EventStore-Client-Go/streamrevision"
+)
+
+type AppendToStreamOptions struct {
+	revision stream_revision.ExpectedRevision
+}
+
+func AppendToStreamOptionsDefault() AppendToStreamOptions {
```

Comment:
Maybe instead of having `AppendToStreamOptionsDefault` it will be nicer to have setDefaults() method that is setting these values if they are missing.

Here is an example how we are doing it in Watermill: https://github.com/ThreeDotsLabs/watermill/blob/e1fb767c92e58ff95d40d56e501ff9966a709426/components/forwarder/publisher.go#L14

and used in the constructor: https://github.com/ThreeDotsLabs/watermill/blob/e1fb767c92e58ff95d40d56e501ff9966a709426/components/forwarder/publisher.go#L36

IMO it's giving the user the best API because they can just pass empty struct and defaults will be set. In that case the user can pas empty struct as well if he will not find that `AppendToStreamOptionsDefault` function exists, and probably it will not work best.

```
pub := forwarder.NewPublisher(forwarder.Config{})
// ...
```

Reply:
My only issue with this approach is it lets the initiated value with wrong values and the user needs to remember it. With the approach I chose, it's impossible for this situation to happen.

Reply:
I'm looking into applying this pattern which I believe address our concerns. What do you think? https://infevo.net/fluent-configure-in-golang/#solution

Reply:
Nooo 😱 "fluent config" is the worst thing in the word :sweat_smile: It's really hard to find what options you can pass.

> My only issue with this approach is it lets the initiated value with wrong values and the user needs to remember it. With the approach I chose, it's impossible for this situation to happen.

Let me take a look at the usages and maybe I'll have something to suggest.

Reply:
Ok, I have something like that in my head:


```go

// options/append.go

type AppendToStreamOptions struct {
	Revision stream_revision.ExpectedRevision
}

// AppendToStreamOptions could be actually moved to client/client.go, so SetDefaults() can be made private.
// I think that it can make logically sense to keep options close to a place where they are use.
func (a *AppendToStreamOptions) SetDefaults() {
	if a.Revision != nil {
		return
	}
	a.Revision = stream_revision.Any()
}

// ...


// client/client.go

func (client *Client) AppendToStream(
	context context.Context,
	streamID string,
	opts options.AppendToStreamOptions,
	events ...messages.ProposedEvent,
) (*WriteResult, error) {
	opts.SetDefaults()


// ...

//
func main() {
	opts := options.AppendToStreamOptions{Revision: stream_revision.NoStream()}
	_, err := client.AppendToStream(context, streamID.String(), opts, testEvent)


	_, err := client.AppendToStream(context, streamID.String(), options.AppendToStreamOptions{}, testEvent)
}

```

It's totally impossible to pass invalid options.AppendToStreamOptions. If `Revision` is empty it will be defaulted to `stream_revision.Any()`. If it isn't value will be just passed.

I think that it may be also worth considering moving `AppendToStreamOptions` to `client/client.go`, so closer to the place where they are used.

Reply:
Ok I run severals ideas locally and I think your idea with the `SetDefault` function is the best. So I'll go that route, along with using in-place setters.

Reply:
after experiencing it locally, moving the `SetDefaults` to private would work. I need to do some work before being able to move those options types in the client package though.

Thanks for the tip!